lookout/style/format/analyzer.py

```diff
@@ -210,6 +212,23 @@ def train(cls, ptr: ReferencePointer, config: Mapping[str, Any], data_request_st
         cls.log.info("trained %s", model)
         return model
 
+    @staticmethod
+    def _predict(X: numpy.ndarray, vnodes_y: Sequence[VirtualNode], vnodes: Sequence[VirtualNode],
+                 rules: Rules, language: str, return_originals: bool = False
+                 ) -> Union[Tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray, numpy.ndarray],
+                            Tuple[numpy.ndarray, numpy.ndarray]]:
+        y_pred, winners = rules.predict(X, True)
```

Comment:
We do this at review time too, so everything should go under `rules.predict`

Reply:
rules has no knowledge of anything related to the nature of what it predicts so to me this seems like it breaks abstraction. WDYT @vmarkovtsev?

Reply:
The expectation is to take a bunch of rules and get the result of their application which you present to the users. How they are applied is completely abstracted away.

Reply:
You can look at the current code and see how twisted it becomes otherwise.

Reply:
BTW I can make predict public in `FormatAnalyzer` to make it clear that this is an integral part of the process that is not a hack.

Reply:
Rename the current `Rules.predict()` to `apply()` (yep, this makes sense, agree) and move `FormatAnalyzer._predict` to `Rules.predict` please.